src/sentry/utils/services.py

```diff
@@ -76,3 +83,173 @@ def expose(self, context):
+        Exception raised when an invalid backend is returned by a selector
+        function.
+        """
+
+    def __init__(self, backend_base, backends, selector_func, callback_func=None):
+        if callback_func is not None:
+            callback_func = import_string(callback_func)
+
+        self.__backend_base = import_string(backend_base)
+
+        def load_executor(options):
+            path = options.get('path')
+            if path is None:
+                executor_cls = ThreadedExecutor
+            else:
+                executor_cls = import_string(path)
+            return executor_cls(**options.get('options', {}))
+
+        self.__backends = {}
+        for name, options in backends.items():
+            self.__backends[name] = (
+                import_string(options['path'])(**options.get('options', {})),
+                load_executor(options.get('executor', {})),
+            )
+
+        self.__selector_func = import_string(selector_func)
+        self.__callback_func = callback_func
+
+    def validate(self):
+        for backend, executor in self.__backends.values():
+            backend.validate()
+
+    def setup(self):
+        for backend, executor in self.__backends.values():
+            backend.setup()
+
+    def __getattr__(self, attribute_name):
+        # When deciding how to handle attribute accesses, we have three
+        # different possible outcomes:
+        # 1. If this is defined as a method on the base implementation, we are
+        #    able delegate it to the backends based on the selector function.
+        # 2. If this is defined as an attribute on the base implementation, we
+        #    are able to (immediately) return that as the value. (This also
+        #    mirrors the behavior of ``LazyServiceWrapper``, which will cache
+        #    any attribute access during ``expose``, so we can't delegate
+        #    attribute access anyway.)
+        # 3. If this isn't defined at all on the base implementation, we let
+        #    the ``AttributeError`` raised by ``getattr`` propagate (mirroring
+        #    normal attribute access behavior for a missing/invalid name.)
+        base_value = getattr(self.__backend_base, attribute_name)
+        if not inspect.ismethod(base_value):
+            return base_value
+
+        def execute(*args, **kwargs):
+            # Binding the call arguments to named arguments has two benefits:
+            # 1. These values always be passed in the same form to the selector
+            #    function and callback, regardless of how they were passed to
+            #    the method itself (as positional arguments, keyword arguments,
+            #    etc.)
+            # 2. This ensures that the given arguments are those supported by
+            #    the base backend itself, which should be a common subset of
+            #    arguments that are supported by all backends.
+            callargs = inspect.getcallargs(base_value, None, *args, **kwargs)
```

Comment:
This requires Python 2.7, I'm not sure whether or not we've ever specified that as a dependency or if I should figure out how to make this work with 2.6? I don't anticipate anyone running this themselves, so as long as we're running 2.7, that might be OK?

Reply:
Ya, seems fine to me. If someone is cooking up `ServiceDelegator` on their own and still running Python 2.6 then I'd like to know wtf they are doing.

Reply:
IIRC Sentry requires 2.7

Reply:
This is correct. We strictly don't support 2.6.

Reply:
Thanks